tests: ztest: add a common fatal and assert hook for special purpose

enjiamai opened the pull request:
This is in order to reduce the redundancy code writing for fatal and assert handler for error case testing. They can be used both in kernel and userspace, and are also SMP safe.

It is target to be part of ZTEST.

Why doing this? For writing error testing case (or we call it negative test case), we might have to write self-defined k_sys_fatal_handler or post_assert_handler to deal with the errors we caught, in order to make test going on. This mean
much identical code would be written. So we try to make it as a common part and let other test case reuse it.  
But when error happened, it sometimes need a special action to make our testing program back to normal, such as release some
resource hold before error happened. This is why we add a hook on it, in order to achieve that goal. 

Usage for dealing with fatal error:
1. Add CONFIG_ZTEST_FATAL_HOOK=y into your prj.conf
2. Include <ztest_fatal_hook.h> in your C code.
3. Define a hook function call ztest_post_fatal_error_hook(), this is optional.
4. Call ztest_set_fault_valid(true) before where your target function call. 

Usage for dealing with assert fail: 
1. Add CONFIG_ZTEST_ASSERT_HOOK=y into your prj.conf
2. Include <ztest_assert_hook.h> in your C code.
3. Define a hook function call ztest_post_assert_fail_hook(), this is optional.
3. call ztest_set_assert_valid(true) before where your target function call.

We split the into 2 induvial CONFIG because it would be more flexible, you can choose to use one or both of them. If you use none of them, you can still defined your own fatal or assert handler as usual.  


Signed-off-by: Enjia Mai <enjiax.mai@intel.com>

Commit:
tests: ztest: add a common fatal and assert hook for special purpose

This is in order to reduce the redundancy code writing for fatal and
assert handler for error case testing. They can be used both in kernel
and userspace, and are also SMP safe.

Signed-off-by: Enjia Mai <enjiax.mai@intel.com>

LeiW000 commented on subsys/testsuite/Kconfig:
```diff
@@ -144,4 +144,22 @@ config TEST_ARM_CORTEX_M
 	  Note that by default, when building with ARM_SECURE_FIRMWARE
 	  set, these exceptions are set to target the Non-Secure state.
 
+config TEST_NEGATIVE_FATAL_HANDLER
```
Does it make sense to use NEGATIVE_TEST_FATAL_HANDLER, rather than TEST_NEGATIVE_FATAL_HANDLER?

enjiamai replied:
Yes, I think so. Updated it already.

LeiW000 commented on subsys/testsuite/Kconfig:
```diff
@@ -144,4 +144,22 @@ config TEST_ARM_CORTEX_M
+	default n
+	help
+	  This option will leverage a pre-defined fatal error handler to deal
+	  with scenario which running the negative test cases. This should be
```
It's better to change the words like:

with scenario, where negative test cases run.

enjiamai replied:
Thanks for correcting this. Got it fixed.

LeiW000 commented on subsys/testsuite/include/negative_handler.h:
```diff
@@ -0,0 +1,58 @@
+/**
+ * @brief Set the flag that treat assert fail happened as expected
+ *
+ * This is used for negative test cases which will trigger a assert fail.
```
Please remove "will", and replace "trigger" with "triggers"

enjiamai replied:
Thanks for correcting this. Got it fixed.

LeiW000 commented on subsys/testsuite/utils/negative_handler.c:
```diff
@@ -0,0 +1,57 @@
+	if (valid_assert) {
+		valid_assert = false; /* reset back to normal */
+	} else {
+		printk("== assert failed was unexpected, aborting ==\n");
```
Pleae change it like:
"Assert failure was not expected. Abording....."

enjiamai replied:
Thanks for correcting this. Got all the relevant places fixed.

enjiamai commented:
@LeiW000 Thanks for reviewing it. I has got all your comments fixed, please help to review it again when you have time~

andyross commented on subsys/testsuite/Kconfig:
```diff
@@ -144,4 +144,23 @@ config TEST_ARM_CORTEX_M
 	  set, these exceptions are set to target the Non-Secure state.
 
+config NEGATIVE_TEST_FATAL_HANDLER
+	bool "Using a pre-defined fatal handler to deal with negative case"
```
Do you really need these to be under a kconfig?  It looks like if you don't call them, none of this code will be linked in.  If that's the case then just make them regular APIs.  You put APIs inside a kconfig if there's non-zero overhead of some form to having them available.

enjiamai replied:
Excuse me, I thought they cannot be regular API, because this is only used for some test cases. And put it under kconfig would make it more flexible, for example, if we want to define a totally different k_sys_fatal_handler with this one, we don't need to do anything. Another advantage is adding the code this way won't influence other test cases in our code base.  That's what I thought, how do you think? thank you so much!

andyross replied:
As I read it, these are just functions, no?  There's no static data with an init hook or anything else that will be automatically linked into the app.  If you don't call the API, then it won't be part of the application.

enjiamai replied:
I tried to make them not under Kconfig but the it will show "multiple definition of k_sys_fatal_handler" during compilations, conflicts with other test case such as mem_protect,  pipe, etc, which already define k_sys_fatal_handler. That why I put them under Kconfig. Excuse me, maybe I have something misunderstanding?

andyross replied:
Ah... OK.  I didn't realize that was a weak symbol and not something registered at runtime.  I guess this is fine, though you could still do this by integrating it with the existing weak fallback version.

andyross commented on subsys/testsuite/include/negative_handler.h:
```diff
@@ -0,0 +1,82 @@
+	 * compiler before we set valid_fault. This can happen with expansion
+	 * of inline syscall invocation functions.
+	 */
+	compiler_barrier();
```
"valid_assert" is already marked volatile.  The comment and compiler barrier are needless, preventing reordering of access is exactly what "volatile" means.  Remove.

enjiamai replied:
Got it, so glad to learn this concept and will study it ! I saw the code in tests/kernel/mem_protect/mem_protect/src/mem_protect.h using it this way, so I refer to it.  I will update them, thanks!

andyross commented on subsys/testsuite/include/negative_handler.h:
```diff
@@ -0,0 +1,82 @@
+	 * compiler before we set valid_fault. This can happen with expansion
+	 * of inline syscall invocation functions.
+	 */
+	compiler_barrier();
```
Same comment, no compiler_barrier() needed for valid_fault, it's already volatile.

enjiamai replied:
Got it, thanks! will get them fixed.

andyross commented on subsys/testsuite/include/negative_handler.h:
```diff
@@ -0,0 +1,82 @@
+ *
+ * @param valid flag indicate fault is expected
+ */
+static inline void set_fault_valid(bool valid)
```
Hygine: this is an API function and part of the ztest suite, so it should be "ztest_set_fault_valid"

andyross commented on subsys/testsuite/include/negative_handler.h:
```diff
@@ -0,0 +1,82 @@
+ * Call this function to pass parameter to predefined fatal handler, it makes
+ * us taking different actions in our test case (applications).
+ */
+static inline void set_fatal_hook_param(void *param)
```
Same comment, APIs should be prefixed with "ztest_"

andyross commented on subsys/testsuite/include/negative_handler.h:
```diff
@@ -0,0 +1,82 @@
+ *
+ * @param valid flag indicate assert is expected
+ */
+static inline void set_assert_valid(bool valid)
```
Same: "ztest_"

enjiamai replied:
Got it, will got it fixed. thanks!

andyross reviewed (changes requested):
Some nitpicking that rises just above the threshold for a -1.

Also, is "negative handler" really a good name for the feature?  That doesn't mean very much to me.  Why not call it "fatal_error_hook" or something, as that is what is actually happening?

andyross commented on subsys/testsuite/include/negative_handler.h:
```diff
@@ -0,0 +1,82 @@
+
+#if defined(CONFIG_NEGATIVE_TEST_FATAL_HANDLER)
+/* These valid flag used to figure out if the fault was expected or not. */
+extern volatile bool valid_fault;
```
It may not be an issue for the specific tests using this feature, but note that this isn't SMP-safe.  A fault happens on a given CPU, calling "set_fault_valid()" will race against anything going wrong on the other CPU (which, again, may not be an issue in practice).

Really this tracking needs to be on the _current_cpu struct, not global.  But again maybe not an issue since it's only a handful of tests that need this and they can arrange things on their own to not have racing faults.

enjiamai replied:
Yes, I totally did not consider this. It indeed be able to cause problem under SMP cases. Thank you so much for your reminder on this! I will take into account this more and think about how to refine this.

enjiamai replied:
I try to give each cpu a valid_fault or valid_assert flag to make sure they can be used in SMP case. But I did not put the in _current_cpu struct. Could you please to see if my method works? thanks!

enjiamai commented:
@andyross I appreciated you review for this and give me so many feedback, I will update them ASAP, thank again!

enjiamai commented:
@andyross I have addressed all the valuable comment you given and could you please help to review it again when you have time? thanks!

andyross reviewed (approved):
Looks fine.  Thanks.

enjiamai commented:
> Looks fine. Thanks.

Thanks again for you spending time on this ^^

andrewboie commented:
Looking this over, I am not sure the per-cpu valid fault logic is robust enough. There is a window in between the invocation of `ztest_set_fault_valid()` and the actual fault being triggered where the caller could get preempted and then rescheduled onto another CPU, in which case the fault would halt the system as it didn't happen on the core it was expected on.

Stepping back a little, do we need per-cpu granularity for this? I guess this was implemented so that we can confirm that the code we expected to fault actually faulted, and not some other system thread that happened to crash at the same time.

I have an alternative idea: how about ztest_set_fault_valid() will do a check:
 - If we are not in an ISR, save the the value of k_current_get() in a global (not per-CPU)
 - If we are in an ISR (irq_offload() use-case) note that in some global boolean

Then in the fatal handler itself, either match the faulting thread (by calling k_current_get() again) or checking k_is_in_isr()

That should help prove that the context we expected to fault actually faulted and not something else.

Other than this SMP concern the PR looks fine.

nashif commented:
Can we please add a user of this, such as a test that utilizes this to show that it works expected and how it works?

nashif reviewed (changes requested):
test needed

enjiamai commented:
> Looking this over, I am not sure the per-cpu valid fault logic is robust enough. There is a window in between the invocation of `ztest_set_fault_valid()` and the actual fault being triggered where the caller could get preempted and then rescheduled onto another CPU, in which case the fault would halt the system as it didn't happen on the core it was expected on.
> 
> Stepping back a little, do we need per-cpu granularity for this? I guess this was implemented so that we can confirm that the code we expected to fault actually faulted, and not some other system thread that happened to crash at the same time.
> 
> I have an alternative idea: how about ztest_set_fault_valid() will do a check:
> 
> * If we are not in an ISR, save the the value of k_current_get() in a global (not per-CPU)
> * If we are in an ISR (irq_offload() use-case) note that in some global boolean
> 
> Then in the fatal handler itself, either match the faulting thread (by calling k_current_get() again) or checking k_is_in_isr(). And clear the saved state information.
> 
> That should help prove that the context we expected to fault actually faulted and not something else.
> 
> Other than this SMP concern the PR looks fine.

@andrewboie I feel this is really a good idea, check if the thread is what we want,  is definitely precise than I did here. Additionally, the justification for the ISR context, a global boolean seem to be robust enough. And I guess by doing this in your way, I may not need to write a syscall for this any more (because I want to get the current cpu id when testing runs in userspace). 
Let me give it a try. Thanks for your suggestion! It is awesome!

enjiamai commented:
> Can we please add a user of this, such as a test that utilizes this to show that it works expected and how it works?

@nashif  Yes, it should have and I am working on it. Thanks for your reminder on this.

Commit:
tests: ztest: add test cases and example for fatal handler hook

Add testing for fatal and assert handler hook feature, and examples
of how to use it.

Signed-off-by: Enjia Mai <enjiax.mai@intel.com>

enjiamai commented:
@nashif I add the test and example, do you think is it OK? thanks!

LeiW000 commented:
@nashif , please take look at the latest code. Development of some test cases depend on it.

ruuddw commented:
The FATAL_ACCESS_NULL 0-pointer test assumes reading from address 0 triggers some exception. However, on certain targets accessing address 0 might actually be valid. See #31419.
Is the test intended to test null pointers, or would testing any illegal memory access be ok? Can it be generalized to use some other address for targets that feature regular access to address 0?

enjiamai commented:
@ruuddw  This test case aim to show we can use this common error handler when fatal error happens. So I think any illegal memory access would be good.

And the identical issues: https://github.com/zephyrproject-rtos/zephyr/issues/31339

I think there are two possible way to fix this:
1. Exclude the FATAL_ACCESS_NULL for ARC-base platform.
2. To access an illegal address which will cause fatal error instead of NULL.

Let me run some trial here. By the way, which one do you prefer?

ruuddw commented:
Thanks, if it is a general memory error than I would prefer 2.
Maybe trying to access a kernel object (I believe this test runs in userspace, right?) or writing to the instruction address region is a more portable/robust way of triggering the error handler.
